File: cpp/spine/systems/system_1_spin.cpp
/***************************************************************
* Author: J.P.G. van Dijk
***************************************************************/

#include "system_1_spin.h"

using namespace spine;

spine::systems::system_1_spin::system_1_spin()
	: spine::systems::system_spin(1)
{
	
}

spine::systems::system_1_spin::~system_1_spin()
{

}

int spine::systems::system_1_spin::getIndex(unsigned int * states)
{
	if (states[0] == STATE_0)
		return 0;
	if (states[0] == STATE_1)
		return 1;
	return -1;
}

vector<unsigned int> * spine::systems::system_1_spin::getIndexMeasurement(unsigned int dot, unsigned int state)
{
	if (dot == 0) {
		if (state == STATE_0)
			return &measurement_0_0;
		if (state == STATE_1)
			return &measurement_0_1;
	}
	return nullptr;
}

unsigned int spine::systems::system_1_spin::getDimension()
{
	return 2;
}

void spine::systems::system_1_spin::updateHamiltonian(realnum * H)
{
	H[0] = -larmor_frequency[0] / ((realnum) 2.0);
	H[1] = rabi_frequency[0] * microwave_control[0];
	H[2] = H[1];
	H[3] = -H[0];
}

File: matlab/+spine/+systems/system_1_spin.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Author: J.P.G. van Dijk                                                 %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef system_1_spin < spine.systems.system_spin

    methods
        
        function obj = system_1_spin()
            obj = obj@spine.systems.system_spin(1);
        end
        
        function index = getIndex(~, states)
            if (states(1) == spine.systems.system_spin.STATE_0)
                index = 1;
            elseif (states(1) == spine.systems.system_spin.STATE_1)
                index = 2;
            end
        end
        
        function index = getIndexMeasurement(~, dot, state)
            index = [];
            if (dot == 1)
                if (state == spine.systems.system_spin.STATE_0)
                    index = 1;
                elseif (state == spine.systems.system_spin.STATE_1)
                    index = 2;
                end
            end
        end
        
        function dimension = getDimension(obj)
            dimension = 2;
        end
        
        function H = updateHamiltonian(obj)
            H = [-obj.larmor_frequency / 2.0, obj.rabi_frequency * obj.microwave_control;
                 obj.rabi_frequency * obj.microwave_control, obj.larmor_frequency / 2.0];
        end

    end
    
end
